Src/FluentAssertions/Common/TypeExtensions.cs

```diff
@@ -136,44 +136,229 @@ public static bool IsEquivalentTo(this IMember property, IMember otherProperty)
 
         public static Type[] GetClosedGenericInterfaces(Type type, Type openGenericType)
```

Comment:
I'm failing to see how this added complexity benefits.
As you mentioned earlier `GetClosedGenericInterfaces` is only used in `GenericDictionaryEquivalencyStep` and `GenericEnumerableEquivalencyStep`.
In both cases `GetClosedGenericInterfaces` is only invoked for `TypeCode.Object`, so `TryGetClosedGenericInterfacesFast` is just a small overhead, as it will return `false` and we continue the slow path.
Am I missing something?

I would recommend you focus on correctness, API and readability/usability in this PR.
It easy to go down the rabbit hole when optimizing (this also points towards myself), and it should be backed up with benchmarks and not just guesses. 

You're always to welcome to open performance related issues/PRs.
I'm happy to discuss ideas on how to avoid Fluent Assertions adding unnecessarily overhead.

Reply:
It gets called on _everything_, because it is used in the implementation of `CanHandle` in `GenericDictionaryEquivalencyStep` and `GenericEnumerableEquivalencyStep`. In fact, because it's in both steps, it gets called twice on every value that passes through the system, including values of primitive types. Every single value being processed by FluentAssertions has its full list of interfaces enumerated at least twice.

This is absolutely necessary for the System.Data code, because the object graphs involve vast numbers of primitive values. I intended to include it in #1434 along with the others, and was dismayed to discover it was still showing as a difference after #1434 was merged. But, #1419 isn't usable without this change.

I am trying to do benchmarking of it but am getting weird errors from BenchmarkDotNet.

Reply:
I made my own benchmark-driving code because for some reason even with a completely empty benchmark class, BenchmarkDotNet is currently throwing an `IndexOutOfRangeException` inside its own `Measurement.Parse` method. My benchmark shows that `GetClosedGenericInterfaces` is a bit over 7 times faster for primitive types with this optimization. To be honest I was expecting more than that, I guess even `Type.GetTypeCode` has some cost to it, but I can say empirically that in actual code using FluentAssertions on `DataTable` objects, this difference is _very_ noticeable.

Reply:
I have spent some hours this morning running your benchmarks and profiling to see if I could find out why introducing `TryGetClosedGenericInterfacesFast` would be faster.
The benchmarks points at `TryGetClosedGenericInterfacesFast` being a de-optimization.
The profiling points at calls to `GetClosedGenericInterfaces` being a negligible part of the total CPU usage.

Can you help me setup some code where you experience it matters?

# Benchmarking
I did some simplifications to the benchmarks:
https://github.com/jnyrup/fluentassertions/commit/5dae0ebe1bafa377f500d124dee88cf0b96d5219

In all cases adding `TryGetClosedGenericInterfacesFast` to `GetCloseGenericInterfaces` was slower than your previous optimizations in `GenericDictionaryEquivalencyStep`/`GenericEnumerableEquivalencyStep` where you checked for `TypeCode.Object` and just returned `Array.Empty<Type>()`.

## UsersOfGetClosedGenericInterfaces

### `TryGetClosedGenericInterfacesFast`
```
|                                     Method |                 Data |      Mean |     Error |    StdDev |
|------------------------------------------- |--------------------- |----------:|----------:|----------:|
| GenericDictionaryEquivalencyStep_CanHandle | Dictionary<int, int> | 650.20 ns | 13.029 ns | 12.797 ns |
| GenericEnumerableEquivalencyStep_CanHandle | Dictionary<int, int> | 646.93 ns |  7.999 ns |  7.482 ns |
| GenericDictionaryEquivalencyStep_CanHandle |            List<int> | 594.56 ns | 11.871 ns | 15.848 ns |
| GenericEnumerableEquivalencyStep_CanHandle |            List<int> | 613.63 ns | 11.134 ns | 10.415 ns |
| GenericDictionaryEquivalencyStep_CanHandle |                 bool |  51.17 ns |  1.023 ns |  1.137 ns |
| GenericEnumerableEquivalencyStep_CanHandle |                 bool |  59.41 ns |  0.930 ns |  0.870 ns |
| GenericDictionaryEquivalencyStep_CanHandle |               string |  54.43 ns |  0.993 ns |  0.928 ns |
| GenericEnumerableEquivalencyStep_CanHandle |               string |  44.61 ns |  0.266 ns |  0.236 ns |
```

### Optimizations in `GenericDictionaryEquivalencyStep`/`GenericEnumerableEquivalencyStep`
```
|                                     Method |                 Data |      Mean |     Error |    StdDev |    Median |
|------------------------------------------- |--------------------- |----------:|----------:|----------:|----------:|
| GenericDictionaryEquivalencyStep_CanHandle | Dictionary<int, int> | 618.10 ns | 12.083 ns | 13.431 ns | 618.34 ns |
| GenericEnumerableEquivalencyStep_CanHandle | Dictionary<int, int> | 623.89 ns | 12.174 ns | 15.830 ns | 629.81 ns |
| GenericDictionaryEquivalencyStep_CanHandle |            List<int> | 550.37 ns | 10.617 ns | 12.227 ns | 543.28 ns |
| GenericEnumerableEquivalencyStep_CanHandle |            List<int> | 609.55 ns | 12.037 ns | 16.477 ns | 612.87 ns |
| GenericDictionaryEquivalencyStep_CanHandle |                 bool |  31.88 ns |  0.564 ns |  0.528 ns |  32.14 ns |
| GenericEnumerableEquivalencyStep_CanHandle |                 bool |  45.57 ns |  0.911 ns |  1.013 ns |  45.93 ns |
| GenericDictionaryEquivalencyStep_CanHandle |               string |  32.63 ns |  0.618 ns |  0.548 ns |  32.86 ns |
| GenericEnumerableEquivalencyStep_CanHandle |               string |  21.39 ns |  0.441 ns |  0.453 ns |  21.71 ns |
```

## LargeDataTableEquivalencyBenchmarks

### `TryGetClosedGenericInterfacesFast`
```
|         Method | RowCount |         Mean |      Error |     StdDev |       Gen 0 |      Gen 1 | Gen 2 | Allocated |
|--------------- |--------- |-------------:|-----------:|-----------:|------------:|-----------:|------:|----------:|
| BeEquivalentTo |       10 |     2.320 ms |  0.0458 ms |  0.0429 ms |    304.6875 |     7.8125 |     - |   1.84 MB |
| BeEquivalentTo |      100 |     7.333 ms |  0.1420 ms |  0.1395 ms |   1070.3125 |    23.4375 |     - |   6.44 MB |
| BeEquivalentTo |     1000 |    53.598 ms |  0.8318 ms |  0.7780 ms |   8600.0000 |   300.0000 |     - |  52.08 MB |
| BeEquivalentTo |    10000 |   532.295 ms |  5.0486 ms |  4.4754 ms |  84000.0000 |  7000.0000 |     - |  508.4 MB |
| BeEquivalentTo |    50000 | 2,762.083 ms | 26.4415 ms | 24.7334 ms | 422000.0000 | 55000.0000 |     - | 2535.8 MB |
```


### Optimizations in Optimizations in `GenericDictionaryEquivalencyStep`/`GenericEnumerableEquivalencyStep`
```
|         Method | RowCount |         Mean |      Error |     StdDev |       Gen 0 |      Gen 1 | Gen 2 |  Allocated |
|--------------- |--------- |-------------:|-----------:|-----------:|------------:|-----------:|------:|-----------:|
| BeEquivalentTo |       10 |     2.295 ms |  0.0318 ms |  0.0281 ms |    304.6875 |     3.9063 |     - |    1.83 MB |
| BeEquivalentTo |      100 |     7.056 ms |  0.1381 ms |  0.1419 ms |   1054.6875 |    23.4375 |     - |    6.34 MB |
| BeEquivalentTo |     1000 |    53.046 ms |  1.0387 ms |  0.9716 ms |   8500.0000 |   300.0000 |     - |   51.41 MB |
| BeEquivalentTo |    10000 |   527.282 ms | 10.0942 ms |  9.4421 ms |  83000.0000 |  6000.0000 |     - |  501.96 MB |
| BeEquivalentTo |    50000 | 2,636.302 ms | 32.3337 ms | 27.0001 ms | 416000.0000 | 54000.0000 |     - | 2503.65 MB |
```

# Profiling

I tried doing a regular profiling of `LargeDataTableEquivalencyBenchmarks` for `RowCount=5_000` and doing ~1000 iterations of `BeEquivalentTo`. 
https://github.com/jnyrup/fluentassertions/commit/3951d93034ec6b359c6e30b3d61efba925e08f98

It gave me the following numbers of Total CPU:
* `GenericDictionaryEquivalencyStep.CanHandle`: 2.5 % (2.15 %)
* `GenericEnumerableEquivalencyStep.CanHandle`: 0.45 % (0.39 %)
* `GetClosedGenericInterfaces`: 0.97 % (0.84 %)

The numbers in parentheses are when scaling to exclude the 14% of CPU time used in initializing `LargeDataTableEquivalencyBenchmarks`

Reply:
The limited optimization currently baked into `GetClosedGenericInterfaces` probably _is_ marginally faster, but in previous communication, I thought we'd established that the fact that that optimization isn't entirely precise and makes assumptions about `TypeCode` values other than `Object` is a problem. So, this restructuring makes it do a little bit more work so that it is accurate across all type codes, explicitly _not_ taking account the external knowledge that it is currently only called by those particular steps.

The optimization started out within `GenericDictionaryEquivalencyStep` and `GenericEnumerableEquivalencyStep`, and migrated from there to be more general in response to previous PR feedback, didn't it?

Reply:
We established that having assumptions in `TypeExtensions.GetClosedGenericInterfaces` about only `TypeCode.Object` being able to implement generic interfaces was not universally true and didn't belong in a general purpose class.
That's why we moved the checks outside of `TypeExtensions` and into `GenericDictionaryEquivalencyStep` and `GenericEnumerableEquivalencyStep` where we have more context, specifically that we always look for either `IDictionary<,>` or `IEnumerable<>` both of which are only implemented by `TypeCode.Object`.
With the exception of `TypeCode.String` implementing `IEnumerable<char>`, but `string` is handled by another `EquivalencyStep`.

I'm absolutely fine about having the checks for `TypeCode.Object` in the two `EquivalencyStep` classes.
I hope that has never been unclear from me.

As `GetClosedGenericInterfaces` is currently only called from the two `EquivalencyStep` classes and with the checks applied inside those classes, `GetClosedGenericInterfaces` will _currently_ always be called for a type having `TypeCode.Object` and `TryGetClosedGenericInterfacesFast` will return `false`.

Having an isolated look `GetClosedGenericInterfaces` `TryGetClosedGenericInterfacesFast` seems to be faster for some cases (I didn't complete the benchmarks for all types), but it doesn't seems to benefit any _current_ cases inside FA.

We should not add code/complexity to optimize for potential future cases.


From the profile I recorded there seems to plenty of other places, where we can squeeze out more performance, both engineering-wise and algorithmically.
For data columns with primitive types, we might only need to find the appropriate `EquivalencyStep` once per column instead of for every cell in that column.

Reply:
I believe you are mistaken about it only ever being called on `TypeCode.Object`, because it is implicated in the `CanHandle` method, not just `Handle` -- that means that it ends up being called on _every single value_ passing through the system, not just the ones that `GenericDictionaryEquivalencyStep` and `GenericEnumerableEquivalencyStep` actually handle. Empirically, I observe a huge performance difference with this optimization, and the reason I did it in the first place was that in earlier iterations, where it was taking at least double digit minutes to compare two moderately sized `DataTable` objects, when I paused execution I often found it enumerating the interfaces implemented by primitive values.

But, also, I clearly misunderstood the intent and I have no problem moving the optimization back so that the two equivalency steps are checking it themselves rather than addressing it in the underlying method. :-)

Reply:
Having the optimizations in `GetEnumerableInterfaces` and `GetDictionaryInterfaces` should cover both `CanHandle` and `Handle`.

Reply:
That is indeed true. :-)

b9505ee

Reply:
This is a good example of why DRY can hurt the maintainability of your code base. Trying to optimize a single instance of some algorithm for every possible situation results in over-engineering. As @jnyrup suggested, having a duplicate implementation next / inside the code that needs that is often much easier to reason about. And performance optimizations must be balanced with maintainability. Adding a lot of complexity may not be worth it if it makes the code base more difficult to understand.